managed/services/dbaas/kubernetes/types.go

```diff
@@ -0,0 +1,373 @@
+func DatabaseClusterForPXC(cluster *dbaasv1beta1.CreatePXCClusterRequest, clusterType ClusterType) (*dbaasv1.DatabaseCluster, error) {
+	if (cluster.Params.Proxysql != nil) == (cluster.Params.Haproxy != nil) {
+		return nil, errors.New("pxc cluster must have one and only one proxy type defined")
+	}
+	memory := cluster.Params.Pxc.ComputeResources.MemoryBytes
+	gCacheSize := "600M"
+	diskSize := resource.NewQuantity(cluster.Params.Pxc.DiskSize, resource.DecimalSI)
+	cpu, err := resource.ParseQuantity(fmt.Sprintf("%dm", cluster.Params.Pxc.ComputeResources.CpuM))
+	if err != nil {
+		return nil, err
+	}
+	clusterMemory := resource.NewQuantity(cluster.Params.Pxc.ComputeResources.MemoryBytes, resource.DecimalSI)
+	if cluster.Params.Pxc.Configuration == "" {
+		if memory > memorySmallSize && memory <= memoryMediumSize {
+			gCacheSize = "2.4G"
+		}
+		if memory > memoryMediumSize && memory <= memoryLargeSize {
+			gCacheSize = "9.6G"
+		}
+		if memory > memoryLargeSize {
+			gCacheSize = "9.6G"
+		}
+		cluster.Params.Pxc.Configuration = fmt.Sprintf(pxcDefaultConfigurationTemplate, gCacheSize)
+	}
+	dbCluster := &dbaasv1.DatabaseCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: cluster.Name,
+		},
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: dbaasAPI,
+			Kind:       dbaasKind,
+		},
+		Spec: dbaasv1.DatabaseSpec{
+			Database:       DatabaseTypePXC,
+			DatabaseImage:  cluster.Params.Pxc.Image,
+			DatabaseConfig: cluster.Params.Pxc.Configuration,
+			ClusterSize:    cluster.Params.ClusterSize,
+			DBInstance: dbaasv1.DBInstanceSpec{
+				DiskSize: *diskSize,
+				CPU:      cpu,
+				Memory:   *clusterMemory,
+			},
+			Monitoring: dbaasv1.MonitoringSpec{
+				PMM: &dbaasv1.PMMSpec{},
+			},
+			LoadBalancer: dbaasv1.LoadBalancerSpec{},
+			Backup:       dbaasv1.BackupSpec{},
+		},
+	}
+	if cluster.Params.Pxc.StorageClass != "" {
+		dbCluster.Spec.DBInstance.StorageClassName = &cluster.Params.Pxc.StorageClass
+	}
+	if cluster.Params.Haproxy != nil {
+		resources, err := convertComputeResource(cluster.Params.Haproxy.ComputeResources)
+		if err != nil {
+			return nil, err
+		}
+		dbCluster.Spec.LoadBalancer.Image = cluster.Params.Haproxy.Image
+		dbCluster.Spec.LoadBalancer.Size = cluster.Params.ClusterSize
+		dbCluster.Spec.LoadBalancer.Resources = resources
+		dbCluster.Spec.LoadBalancer.Type = "haproxy"
+		dbCluster.Spec.LoadBalancer.Configuration = haProxyDefaultConfigurationTemplate
+		dbCluster.Spec.LoadBalancer.TrafficPolicy = "Cluster"
+	}
+	if cluster.Params.Proxysql != nil {
+		resources, err := convertComputeResource(cluster.Params.Proxysql.ComputeResources)
+		if err != nil {
+			return nil, err
+		}
+		dbCluster.Spec.LoadBalancer.Size = cluster.Params.ClusterSize
+		dbCluster.Spec.LoadBalancer.Image = cluster.Params.Proxysql.Image
+		dbCluster.Spec.LoadBalancer.Resources = resources
+		dbCluster.Spec.LoadBalancer.Type = "proxysql"
+	}
+	if cluster.Expose {
```

Comment:
What if `Expose` is true but `Haproxy` **and** `Proxysql` are both nil? Will expose work?
What if `Haproxy` and `Proxysql` both have values?